### PR TITLE
Deploy --function deploys function

### LINKED ISSUE #3204: UX Improvement: Make deploy function easier
Can't tell you how many times I type  `sls deploy -f functionName` and forget the word function... and then have to wait for the whole stack to update.

Can we alias `sls deploy -f functionName` to `sls deploy function -f functionName` as a quick UX win?

So if a user types `sls deploy -f lolCatz` it will only deply the `lolCatz` function?

### PR DESCRIPTION
**Pull request by HyperBrain**
## What did you implement:

Closes #3204 

## How did you implement it:

* `pluginManager.spawn()` now supports an option `terminateLifecycleAfterExecution`.
When set, the *current* lifecycle will be terminated after the spawned lifecycle returns. PluginManager will automatically continue with the *outer* lifecycle, in case we were spawned too. This allows implementation of commands that spawn other commands, depending on the given Serverless options. A terminating spawn should be used in the very first lifecycle event of the *current* lifecycle, to make sure that hooks into it are not yet executed. Using it in `before:<-- delegating command -->` is sufficient to give this guarantee.
* `deploy:deploy` now checks for a function parameter
If found, it executes a terminating spawn of `deploy:function` in `before:deploy:deploy` which effectively starts the function deploy command and terminates the deploy command, before it started. This works even, if `deploy:deploy` has been spawned from another command (see above).
* Added debug logging (SLS_DEBUG set) on command invocation and explicit termination
This helps in general to track the command chain (run() and spawn()) in case of errors. Also of great value for plugin authors.

## How can we verify it:

Run `serverless deploy` and `serverless deploy --function=XXXX`. The result should look like this:

```
$ node node_modules/serverless/bin/serverless deploy
Serverless: Bundling with Webpack...
[...]
Serverless: Packaging service...
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
.....
Serverless: Stack create finished...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
................................................
Serverless: Stack update finished...
Service Information
service: babel-dynamically-entries-example
stage: dev
region: us-east-1
stack: babel-dynamically-entries-example-dev
api keys:
  None
endpoints:
  GET - https://****.execute-api.us-east-1.amazonaws.com/dev/first
  GET - https://****.execute-api.us-east-1.amazonaws.com/dev/second
functions:
  first: babel-dynamically-entries-example-dev-first
  second: babel-dynamically-entries-example-dev-second
```

```
$ node node_modules/serverless/bin/serverless deploy --function=first
Serverless: Bundling with Webpack...
[...]
Serverless: Packaging function: first...
Serverless: Uploading function: first (71.11 KB)...
Serverless: Successfully deployed function: first
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/classes/PluginManager.js [hunk 1]
```diff
@@ -9,6 +9,21 @@ const getCacheFilePath = require('../utils/getCacheFilePath');
...
+ * @private
+ * Error type to terminate the currently running hook chain successfully without
+ * executing the rest of the current command's lifecycle chain.
+ */
+class TerminateHookChain extends Error {
+  constructor(commands) {
+    const commandChain = _.join(commands, ':');
+
+    super(commandChain);
+    this.message = commandChain;
```
Comment (HyperBrain) [inline lib/classes/PluginManager.js]: @pmuens Message could also be `Terminating ${commandChain}`, although this exception is only a control structure and never will lead to an exception on the outside.

**Comment by HyperBrain** [comment]
Squashed the commits and added docs and unit tests - ready for review/merge now

**Comment by pmuens** [comment]
> Squashed the commits and added docs and unit tests - ready for review/merge now

Great! Thanks for the update and your work on this @HyperBrain 👍. We'll review it and test it ASAP.

**Comment by HyperBrain** [comment]
Ok, will check right now what's wrong with the tests.

**Comment by HyperBrain** [comment]
@pmuens Done. The problem was, that the AWS info unit test did not set a valid `serverless.cli` object, and every code that tried to log something crashed (now pluginManager.invoke() logs when SLS_DEBUG is set).

BTW: I'm not sure if the unit tests should run with SLS_DEBUG set at all.

**Comment by eahefnawy** [comment]
@HyperBrain thanks a lot!! quick question though, I don't see you passing the options to the spawn anywhere. Maybe it's not needed thanks to how spawn works, but just wanna make sure that if we `sls deploy -f hello -s beta`, it'd really spawn `sls deploy function -f hello -s beta` not just `sls deploy function -f hello`.

**Comment by HyperBrain** [comment]
@eahefnawy `spawn()` does not need any options, because it just runs the given command in the existing context, so serverless.options is exactly the same. It does not "spawn" a new process, but only the command.

**Comment by HyperBrain** [comment]
@pmuens I'm not sure if the one failing test has anything to do with my implementation.

Travis:
`1) Create #create() should generate scaffolding for "aws-scala-sbt" template:`

Do you have any idea?

**Comment by pmuens** [comment]
Thanks for getting back and updating @HyperBrain 👍

> @pmuens I'm not sure if the one failing test has anything to do with my implementation.
> Travis:
> 1) Create #create() should generate scaffolding for "aws-scala-sbt" template:
> Do you have any idea?

That's a randomly failing test (I saw that multiple times in the last couple of months). Will look into it this week. I'll restart the build...

**Comment by eahefnawy** [comment]
Tested it today and had no issue. Let's merge!

Thanks again @HyperBrain ... as always!! ❤️

**Pull request merged**